service/history/workflow/update/update.go

```diff
@@ -186,6 +190,53 @@ func (u *Update) WaitAccepted(ctx context.Context) (*updatepb.Outcome, error) {
 	return nil, nil
 }
 
+// WaitLifecycleStage waits until the Update has reached Completed or a timeout.
+// If the Update reaches Completed with no timeout, the outcome is returned. If
+// there is a timeout due to the supplied server timeout, then a nil outcome is
+// returned, without an error. If there is a timeout due to context deadline
+// expiry, then the error is returned.
+func (u *Update) WaitLifecycleStage(ctx context.Context, waitStage enumspb.UpdateWorkflowExecutionLifecycleStage, serverTimeout time.Duration) (*updatepb.Outcome, error) {
+	switch waitStage {
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED:
+		return u.waitLifecycleStage(ctx, u.WaitAccepted, serverTimeout)
+	case enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED:
+		return u.waitLifecycleStage(ctx, u.WaitOutcome, serverTimeout)
+	default:
+		return nil, serviceerror.NewUnimplemented(
+			fmt.Sprintf("%v is not implemented", waitStage))
+	}
+}
+
+func (u *Update) waitLifecycleStage(ctx context.Context, waitFn func(ctx context.Context) (*updatepb.Outcome, error), serverTimeout time.Duration) (outcome *updatepb.Outcome, err error) {
+	ctxDeadline, hasCtxDeadline := ctx.Deadline()
+	serverDeadline := time.Now().Add(serverTimeout)
+	if !hasCtxDeadline || ctxDeadline.After(serverDeadline) {
```

Comment:
```go
pollContext, cancel := context.WithTimeout(context.Background(), serverTimeout)
defer cancel()
type result struct{
  outcome *updatepb.Outcome
  err error
}
resultCh := make(chan struct{}, 1)
go func() {
  outcome, err := waitFn(pollContext)
  resultCh <- result{outcome, err}
}()

select {
  case res <-resultCh:
    if pollContext.Err() {
      // return empty
    }
    return res.outcome, res.err
  case <-ctx.Done():
    return ctx.Err()
}
```

Reply:
That's what I'd do given the signature of `waitFn`

Reply:
Thanks @bergundy, that's cleaner. The PR contains a version derived from that.